### ISSUE TITLE
Add a way to complete an unfinished build without restarting it

### ISSUE THREAD

**Issue by cooljeanius** [fink/fink#221]
In fink/fink-distributions/pull/656 I was trying to debug a failing FPC build, and eventually we found a way to allow the build to succeed manually, but unfortunately there didn't seem to be a way to translate this manual success into a success from within fink's own ways of doing things. @kamischi said:

> Hm. Sorry, but I don't know of a way to complete the build without restarting it. 

Such a feature would be useful for situations like these.

**Comment by dmacks** [comment]
I thought we had a looong-ago feature-request for this. It's not hard to roll a .deb manually. But it's not a feature I'd want to see deployed in fink itself because it contradicts the standard that the .deb represents what the .info gives. It's no problem for personal use, but it's a huge problem when users might later ask for support using this package or any other package that has it as a dependency, because we'd never be able to reproduce it or understand what the state of their system was.

**Comment by cooljeanius** [comment]
Just noting that the fact that this doesn't exist is part of the reason why I opened #220; there's not really much reason for keeping around unfinished builds if there isn't a way to complete them from within `fink`.

**Comment by nieder** [comment]
Unfinished build trees are useful for diagnostics (checking config.log, etc).

But something like this maybe:
```perl
sub cleanup_buildtrees {
	my %opts = (dryrun => 0, @_);

	my $srcdir = "$basepath/src";
	my $build_path = $config->param_default("buildpath", "$basepath/src/fink.build");

	# Get the list of directories inside $build_path that will be removed.
	opendir(DIR, $build_path) or die "Can't access $build_path: $!";
	my @old_buildtrees = readdir $build_path;
	closedir(DIR);

	my $verb;
	if ($opts{dryrun}) {
		$verb = 'Obsolete';
	} else {
		$verb = 'Removing obsolete';
	}

	foreach my $directory (sort @old_buildtrees) {
	# This will kill an active build process.
		print "$verb sourcetree: $srcdir/$directory\n" if $print_it;
		if (!$opts{dryrun}) {
			unlink "$srcdir/$directory" and $file_count++;
		}
	}
}
```

There's a comment earlier in the code about adding a `--dirs` flag to `fink cleanup --srcs`, but I think it should be a separate command (might want to keep obsolete tarballs but still cleanup failed trees).

**Comment by cooljeanius** [comment]
> Unfinished build trees are useful for diagnostics (checking config.log, etc).
> 

Yeah, and they'd be even more useful if we could actually complete their builds.

> But something like this maybe:
> 
> ```perl
> sub cleanup_buildtrees {
> 	my %opts = (dryrun => 0, @_);
> 
> 	my $srcdir = "$basepath/src";
> 	my $build_path = $config->param_default("Buildpath", "$basepath/src/fink.build");
> 
> 	my $file_count = 0;
> 
> 	# Get the list of directories inside $build_path that will be removed.
> 	opendir(DIR, $build_path) or die "Can't access $build_path: $!";
> 	my @old_buildtrees = readdir(DIR);
> 	closedir(DIR);
> 
> 	my $print_it = $opts{dryrun} || $config->verbosity_level() > 1;
> 
> 	my $verb;
> 	if ($opts{dryrun}) {
> 		$verb = 'Obsolete';
> 	} else {
> 		$verb = 'Removing obsolete';
> 	}
> 
> 	foreach my $directory (sort @old_buildtrees) {
> 		if (-d "$build_path/$directory" && $directory ne q(.) && $directory ne q(..)) {
> 			# This will kill an active build process.
> 			print "$verb source tree: $build_path/$directory\n" if $print_it;
> 			if (!$opts{dryrun}) {
> 				unlink "$build_path/$directory" and $file_count++;
> 			}
> 		}
> 	}
> }
> ```
> 
> There's a comment earlier in the code about adding a `--dirs` flag to `fink cleanup --srcs`, but I think it should be a separate command (might want to keep obsolete tarballs but still cleanup failed trees).

This (and PR #258) address the cleaning portion that I opened #220 for, but I don't think it's relevant to the part about finishing builds that this one is about?
